python_modules/libraries/dagster-k8s/dagster_k8s/pipes.py

```diff
@@ -303,24 +305,54 @@ def build_pod_body(
             "A restart policy of Always is not allowed, computations are expected to complete."
         )
 
+    containers = spec["containers"]
+    init_containers = spec.get("init_containers") or []
+
     if "image" not in spec["containers"][0] and not image:
         raise DagsterInvariantViolationError(
             "Must specify image property or provide base_pod_spec with one set."
         )
 
-    if "name" not in spec["containers"][0]:
-        spec["containers"][0]["name"] = DEFAULT_CONTAINER_NAME
+    # We set the container name for the first container in the list if it is not set.
+    # There will be a validation error below for other containers.
+    if "name" not in containers[0]:
+        containers[0]["name"] = DEFAULT_CONTAINER_NAME
+
+    if not init_containers and len(containers) == 1:
+        if image:
+            containers[0]["image"] = image
+
+        if command:
+            containers[0]["command"] = command
+    else:
+        if image:
+            raise DagsterInvariantViolationError(
+                "Should specify 'image' via 'base_pod_spec' when specifying multiple containers"
+            )
+
+        if command:
+            raise DagsterInvariantViolationError(
+                "Should specify 'command' via 'base_pod_spec' when specifying multiple containers"
+            )
 
-    if image:
-        spec["containers"][0]["image"] = image
+        for container in (containers + init_containers):
+            for key in ["name", "image"]:
+                if key not in container:
+                    raise DagsterInvariantViolationError(f"Must provide base_pod_spec with {key} property set.")
 
-    if command:
-        spec["containers"][0]["command"] = command
+    if "env" not in containers[0]:
+        containers[0]["env"] = []
 
-    if "env" not in spec["containers"][0]:
-        spec["containers"][0]["env"] = []
+    # Substitute the Dagster session variables in the pod spec
+    injected_pipes_session_vars = False
+    for container in (containers + init_containers):
+        for env in container.get("env", []):
+            if env["name"] in [DAGSTER_PIPES_CONTEXT_ENV_VAR, DAGSTER_PIPES_MESSAGES_ENV_VAR]:
+                injected_pipes_session_vars = True
+                env["value"] = env_vars[env["name"]]
```

Comment:
should probably call this out in a doc block somewhere, maybe as a note in the `base_pod_spec` arg on `run`?